src/server/svr_jobfunc.c

```diff
@@ -5290,6 +5297,60 @@ svr_clean_job_history(struct work_task *pwt)
 		}
 		/* restore the saved next in pjob */
 		pjob = nxpjob;
+
+		/* check if we spent too long hogging the pbs_server process here */
+		end_time = time(NULL);
+		if ((end_time - begin_time) > SVR_CLEAN_JOBHIST_SECS) {
+			/* Apparently the interval between history purges is too long */
+			/* reduce it using factor 0.7                                 */
+			time_between_tasks = (floor((double)time_between_tasks * 0.7));
+
+			/* no use reducing to less than 4 * SVR_CLEAN_JOBHIST_SECS */
+			/* since we'll already schedule a continuation task here */
+			if (time_between_tasks < (4 * SVR_CLEAN_JOBHIST_SECS))
+				time_between_tasks = 4 * SVR_CLEAN_JOBHIST_SECS;
+
+			/* set up another work task in near future, */
+			/* but leave as much time as we spent in this routine for other work first */
+			if (!set_task(WORK_Timed,
+				(end_time + SVR_CLEAN_JOBHIST_SECS),
+				svr_clean_job_history, (void *)0)) {
+				log_err(errno,
+					"svr_clean_job_history",
+					"Unable to set task for clean job history");
+					/* on error to set task */
+					/* just continue purging the history */
+			} else
+				/* but if we managed to set a task in near future, return; */
+				/* that task will continue where we left off */
+				return;
+		}
+	} /* end of while loop through jobs */
+
+	/* We purged everything necessary in this task if we get here */
+
+	/* set up another work task for next time period */
```

Comment:
Can we have this set_task() call below, after manipulation of time_between_tasks?

Reply:
No. The logic below is trying to move the time up and that should be done after the next task is set.